devito/dle/backends/advanced.py

```diff
@@ -0,0 +1,434 @@
+
+from collections import OrderedDict, namedtuple
+from itertools import combinations
+
+import numpy as np
+import psutil
+
+import cgen as c
+
+from devito.dimension import Dimension
+from devito.dle import (compose_nodes, copy_arrays, filter_iterations,
+                        retrieve_iteration_tree)
+from devito.dle.backends import (BasicRewriter, BlockingArg, dle_pass, omplang,
+                                 simdinfo, get_simd_flag, get_simd_items)
+from devito.dse import promote_scalar_expressions
+from devito.exceptions import DLEException
+from devito.interfaces import TensorFunction
+from devito.logger import dle_warning
+from devito.nodes import Block, Denormals, Element, Expression, Iteration, List
+from devito.tools import as_tuple, flatten, grouper, roundm
+from devito.visitors import (FindNodes, FindSymbols, IsPerfectIteration,
+                             SubstituteExpression, Transformer)
+
+
+class DevitoRewriter(BasicRewriter):
+
+    def _pipeline(self, state):
+        self._avoid_denormals(state)
+        self._loop_fission(state)
+        self._create_elemental_functions(state)
+        self._loop_blocking(state)
+        self._simdize(state)
+        if self.params.get('openmp') is True:
+            self._ompize(state)
+
+    @dle_pass
+    def _loop_fission(self, state, **kwargs):
+        """
+        Apply loop fission to innermost :class:`Iteration` objects. This pass
+        is not applied if the number of statements in an Iteration's body is
+        lower than ``self.thresholds['fission'].``
+        """
+
+        processed = []
+        for node in state.nodes:
+            mapper = {}
+            for tree in retrieve_iteration_tree(node):
+                if len(tree) <= 1:
+                    # Heuristically avoided
+                    continue
+
+                candidate = tree[-1]
+                expressions = [e for e in candidate.nodes if e.is_Expression]
+
+                if len(expressions) < self.thresholds['max_fission']:
+                    # Heuristically avoided
+                    continue
+                if len(expressions) != len(candidate.nodes):
+                    # Dangerous for correctness
+                    continue
+
+                functions = list(set.union(*[set(e.functions) for e in expressions]))
+                stencils = [e.stencil for e in expressions]
+
+                if not functions or not stencils:
+                    # Heuristically avoided
+                    continue
+
+                # Promote temporaries from scalar to tensors
+                handle = functions[0]
+                dim = handle.indices[-1]
+                size = handle.shape[-1]
+                if any(dim != i.indices[-1] for i in functions):
+                    # Dangerous for correctness
+                    continue
+
+                stencils = promote_scalar_expressions(stencils, (size,), (dim,), True)
+
+                assert len(stencils) == len(expressions)
```

Comment:
Asserts are ignored in python's optimised mode. Plus you want to inform the application code that something went wrong - which might not always be the same as stopping execution.